### DISCUSSION TITLE
loss function of cox model

### DISCUSSION THREAD

**Discussion by EQUIWDH** [pbreheny/ncvreg#43] [Q&A] [answered]
Hello, I am wondering if `fit$loss` returned by `ncvsurv` represents only the cox -2*logL calculating by estImated $\beta$ or the -2logL + penalty form ，thanks!

**Comment by EQUIWDH** [comment]
Sorry for disturbing, I am using the loss function structure provided by `loss.R` which is
```R
loss<- function(x,beta,y ,total=TRUE) {
  eta = x%*%beta
  ind <- order(y[,1])
  d <- as.double(y[ind,2])
  if (is.matrix(eta)) {
    eta <- eta[ind, , drop=FALSE]
    r <- apply(eta, 2, function(x) rev(cumsum(rev(exp(x)))))
  } else {
    eta <- eta[ind]
    r <- rev(cumsum(rev(exp(eta))))
  }
  if (total) {
    return(-2*(crossprod(d, eta) - crossprod(d, log(r))))
  } else { 
    return(-2*(eta[d==1,] - log(r)[d==1,]))
  }
}
```
compare`loss(fit$X,fit$beta,train$y)` and `fit$loss`, I find they don't correspond as the picture showing

**Comment by EQUIWDH** [comment]
![test](https://user-images.githubusercontent.com/101512154/233275067-bbb4fcc4-7352-40c1-9cf4-21b27fe14bbe.png)

**Reply by pbreheny** [reply, marked as answer]
`fit$loss` returns the loss only (negative of cox partial likelihood), not the penalty.

The internal function `ncvreg:::loss.ncvsurv()`, as well as the modified version of it that you show above, assumes that the data has been ordered by survival time. In addition, `loss.ncvsurv()` returns the deviance (2*loss).

So:

```r
library(ncvreg)
data(Lung)
X <- Lung$X
y <- Lung$y
fit <- ncvsurv(X, y)
l1 <- 2*fit$loss
l2 <- ncvreg:::loss.ncvsurv(cbind(fit$time, fit$fail), fit$linear.predictors)
cbind(l1, drop(l2)) |> head()
#>                l1          
#> 0.44806 1010.6732 1010.6732
#> 0.41786 1000.5671 1000.5555
#> 0.38970  992.8519  992.8484
#> 0.36344  986.8651  986.8599
#> 0.33894  982.1560  982.1534
#> 0.31610  978.4471  978.4455
```

**Reply by EQUIWDH** [reply]
ok，understand，but if I use AIC function provided in stats package. The IC form will be -2*log(partial likelihood) + 2*(parameter number) right?

**Comment by pbreheny** [comment]
Correct
